e2e/test/scenarios/native/reproductions/15946-mongo-pre-select-table.cy.spec.js

```diff
@@ -17,10 +17,13 @@ describe.skip("issue 15946", { tags: "@external" }, () => {
   it("converting a question to the native query should pre-select a table (metabase#15946)", () => {
     cy.get(".QueryBuilder .Icon-sql").click();
 
-    modal().findByText("Convert this question to a native query").click();
-    modal().should("not.exist");
+    cy.get(".Modal")
+      .findByText("Convert this question to a native query")
+      .click();
+    cy.get(".Modal").should("not.exist");
 
     cy.get(".GuiBuilder-data").contains(MONGO_DB_NAME);
     cy.get(".GuiBuilder-data").contains("Orders");
+    cy.get("aside .RunButton").should("not.be.disabled");
```

Comment:
Could you instead run the query and make sure it works? I think there should be a helper for that

Reply:
> Could you instead run the query and make sure it works? 

Running the query fails in the backend (both UI and tests):

POST http://localhost:3000/api/dataset gives HTTP 202 and this response:

<details>

<summary>Response</summary>

```json
{
	"database_id": 21,
	"started_at": "2023-05-29T17:02:02.162565347+07:00",
	"via": [{
		"status": "failed",
		"class": "class clojure.lang.ExceptionInfo",
		"error": "Unable to parse query: readStartArray can only be called when CurrentBSONType is ARRAY, not when CurrentBSONType is STRING.",
		"stacktrace": ["--> driver.mongo.query_processor$parse_query_string.invokeStatic(query_processor.clj:1334)", "driver.mongo.query_processor$parse_query_string.invoke(query_processor.clj:1327)", "driver.mongo.execute$execute_reducible_query.invokeStatic(execute.clj:177)", "driver.mongo.execute$execute_reducible_query.invoke(execute.clj:173)", "driver.mongo$fn__127630$f__126182__auto____127631.invoke(mongo.clj:312)", "driver.mongo.util$do_with_mongo_connection$fn__126170.invoke(util.clj:230)", "util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)", "util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)", "driver.mongo.util$do_with_mongo_connection.invokeStatic(util.clj:224)", "driver.mongo.util$do_with_mongo_connection.invoke(util.clj:219)", "driver.mongo$fn__127630.invokeStatic(mongo.clj:311)", "driver.mongo$fn__127630.invoke(mongo.clj:309)", "query_processor.context$executef.invokeStatic(context.clj:60)", "query_processor.context$executef.invoke(context.clj:49)", "query_processor.context.default$default_runf.invokeStatic(default.clj:68)", "query_processor.context.default$default_runf.invoke(default.clj:66)", "query_processor.context$runf.invokeStatic(context.clj:46)", "query_processor.context$runf.invoke(context.clj:40)", "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:15)", "query_processor.reducible$identity_qp.invoke(reducible.clj:12)", "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___90965.invoke(cache.clj:224)", "query_processor.middleware.permissions$check_query_permissions$fn__91305.invoke(permissions.clj:126)", "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__92777.invoke(mbql_to_native.clj:24)", "query_processor$fn__95218$combined_post_process__95223$combined_post_process_STAR___95224.invoke(query_processor.clj:260)", "query_processor$fn__95218$combined_pre_process__95219$combined_pre_process_STAR___95220.invoke(query_processor.clj:257)", "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__94047$fn__94052.invoke(resolve_database_and_driver.clj:36)", "driver$do_with_driver.invokeStatic(driver.clj:90)", "driver$do_with_driver.invoke(driver.clj:86)", "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__94047.invoke(resolve_database_and_driver.clj:35)", "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__92580.invoke(fetch_source_query.clj:310)", "query_processor.middleware.store$initialize_store$fn__95033$fn__95034.invoke(store.clj:12)", "query_processor.store$do_with_store.invokeStatic(store.clj:46)", "query_processor.store$do_with_store.invoke(store.clj:40)", "query_processor.middleware.store$initialize_store$fn__95033.invoke(store.clj:11)", "query_processor.middleware.normalize_query$normalize$fn__92807.invoke(normalize_query.clj:36)", "query_processor.middleware.constraints$add_default_userland_constraints$fn__89779.invoke(constraints.clj:54)", "query_processor.middleware.process_userland_query$process_userland_query$fn__93919.invoke(process_userland_query.clj:151)", "query_processor.middleware.catch_exceptions$catch_exceptions$fn__91429.invoke(catch_exceptions.clj:171)", "query_processor.reducible$async_qp$qp_STAR___79973$thunk__79975.invoke(reducible.clj:103)", "query_processor.reducible$async_qp$qp_STAR___79973.invoke(reducible.clj:109)", "query_processor.reducible$async_qp$qp_STAR___79973.invoke(reducible.clj:94)", "query_processor$base_qp$fn__95230.doInvoke(query_processor.clj:278)", "query_processor.reducible$sync_qp$qp_STAR___79985.doInvoke(reducible.clj:129)", "query_processor$process_userland_query.invokeStatic(query_processor.clj:383)", "query_processor$process_userland_query.doInvoke(query_processor.clj:379)", "query_processor$eval95267$process_query_and_save_execution_BANG___95276$fn__95279.invoke(query_processor.clj:394)", "query_processor$eval95267$process_query_and_save_execution_BANG___95276.invoke(query_processor.clj:387)", "query_processor$eval95312$process_query_and_save_with_max_results_constraints_BANG___95321$fn__95324.invoke(query_processor.clj:406)", "query_processor$eval95312$process_query_and_save_with_max_results_constraints_BANG___95321.invoke(query_processor.clj:399)", "api.dataset$run_query_async$fn__113755.invoke(dataset.clj:74)", "query_processor.streaming$streaming_response_STAR_$fn__99869$fn__99870.invoke(streaming.clj:166)", "query_processor.streaming$streaming_response_STAR_$fn__99869.invoke(streaming.clj:165)", "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:69)", "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:67)", "async.streaming_response$do_f_async$task__28914.invoke(streaming_response.clj:88)"],
		"error_type": "invalid-query",
		"ex-data": {
			"type": "invalid-query",
			"query": "\"[{\\\"$project\\\" #ordered/map ([\\\"_id\\\" \\\"$_id\\\"] [\\\"id\\\" \\\"$id\\\"] [\\\"user_id\\\" \\\"$user_id\\\"] [\\\"product_id\\\" \\\"$product_id\\\"] [\\\"subtotal\\\" \\\"$subtotal\\\"] [\\\"tax\\\" \\\"$tax\\\"] [\\\"total\\\" \\\"$total\\\"] [\\\"created_at\\\" \\\"$created_at\\\"] [\\\"quantity\\\" \\\"$quantity\\\"] [\\\"discount\\\" \\\"$discount\\\"])} {\\\"$limit\\\" 1048575}]\""
		}
	}],
	"error_type": "invalid-query",
	"json_query": {
		"database": 21,
		"native": {
			"template-tags": {},
			"query": "\"[{\\\"$project\\\" #ordered/map ([\\\"_id\\\" \\\"$_id\\\"] [\\\"id\\\" \\\"$id\\\"] [\\\"user_id\\\" \\\"$user_id\\\"] [\\\"product_id\\\" \\\"$product_id\\\"] [\\\"subtotal\\\" \\\"$subtotal\\\"] [\\\"tax\\\" \\\"$tax\\\"] [\\\"total\\\" \\\"$total\\\"] [\\\"created_at\\\" \\\"$created_at\\\"] [\\\"quantity\\\" \\\"$quantity\\\"] [\\\"discount\\\" \\\"$discount\\\"])} {\\\"$limit\\\" 1048575}]\"",
			"collection": "orders"
		},
		"type": "native",
		"parameters": [],
		"middleware": {
			"js-int-to-string?": true,
			"add-default-userland-constraints?": true
		}
	},
	"status": "failed",
	"class": "class org.bson.BsonInvalidOperationException",
	"stacktrace": ["org.bson.AbstractBsonReader.verifyBSONType(AbstractBsonReader.java:690)", "org.bson.AbstractBsonReader.checkPreconditions(AbstractBsonReader.java:722)", "org.bson.AbstractBsonReader.readStartArray(AbstractBsonReader.java:443)", "org.bson.codecs.BsonArrayCodec.decode(BsonArrayCodec.java:63)", "org.bson.BsonArray.parse(BsonArray.java:75)", "--> driver.mongo.query_processor$parse_query_string.invokeStatic(query_processor.clj:1332)", "driver.mongo.query_processor$parse_query_string.invoke(query_processor.clj:1327)", "driver.mongo.execute$execute_reducible_query.invokeStatic(execute.clj:177)", "driver.mongo.execute$execute_reducible_query.invoke(execute.clj:173)", "driver.mongo$fn__127630$f__126182__auto____127631.invoke(mongo.clj:312)", "driver.mongo.util$do_with_mongo_connection$fn__126170.invoke(util.clj:230)", "util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:141)", "util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:130)", "driver.mongo.util$do_with_mongo_connection.invokeStatic(util.clj:224)", "driver.mongo.util$do_with_mongo_connection.invoke(util.clj:219)", "driver.mongo$fn__127630.invokeStatic(mongo.clj:311)", "driver.mongo$fn__127630.invoke(mongo.clj:309)", "query_processor.context$executef.invokeStatic(context.clj:60)", "query_processor.context$executef.invoke(context.clj:49)", "query_processor.context.default$default_runf.invokeStatic(default.clj:68)", "query_processor.context.default$default_runf.invoke(default.clj:66)", "query_processor.context$runf.invokeStatic(context.clj:46)", "query_processor.context$runf.invoke(context.clj:40)", "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:15)", "query_processor.reducible$identity_qp.invoke(reducible.clj:12)", "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___90965.invoke(cache.clj:224)", "query_processor.middleware.permissions$check_query_permissions$fn__91305.invoke(permissions.clj:126)", "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__92777.invoke(mbql_to_native.clj:24)", "query_processor$fn__95218$combined_post_process__95223$combined_post_process_STAR___95224.invoke(query_processor.clj:260)", "query_processor$fn__95218$combined_pre_process__95219$combined_pre_process_STAR___95220.invoke(query_processor.clj:257)", "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__94047$fn__94052.invoke(resolve_database_and_driver.clj:36)", "driver$do_with_driver.invokeStatic(driver.clj:90)", "driver$do_with_driver.invoke(driver.clj:86)", "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__94047.invoke(resolve_database_and_driver.clj:35)", "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__92580.invoke(fetch_source_query.clj:310)", "query_processor.middleware.store$initialize_store$fn__95033$fn__95034.invoke(store.clj:12)", "query_processor.store$do_with_store.invokeStatic(store.clj:46)", "query_processor.store$do_with_store.invoke(store.clj:40)", "query_processor.middleware.store$initialize_store$fn__95033.invoke(store.clj:11)", "query_processor.middleware.normalize_query$normalize$fn__92807.invoke(normalize_query.clj:36)", "query_processor.middleware.constraints$add_default_userland_constraints$fn__89779.invoke(constraints.clj:54)", "query_processor.middleware.process_userland_query$process_userland_query$fn__93919.invoke(process_userland_query.clj:151)", "query_processor.middleware.catch_exceptions$catch_exceptions$fn__91429.invoke(catch_exceptions.clj:171)", "query_processor.reducible$async_qp$qp_STAR___79973$thunk__79975.invoke(reducible.clj:103)", "query_processor.reducible$async_qp$qp_STAR___79973.invoke(reducible.clj:109)", "query_processor.reducible$async_qp$qp_STAR___79973.invoke(reducible.clj:94)", "query_processor$base_qp$fn__95230.doInvoke(query_processor.clj:278)", "query_processor.reducible$sync_qp$qp_STAR___79985.doInvoke(reducible.clj:129)", "query_processor$process_userland_query.invokeStatic(query_processor.clj:383)", "query_processor$process_userland_query.doInvoke(query_processor.clj:379)", "query_processor$eval95267$process_query_and_save_execution_BANG___95276$fn__95279.invoke(query_processor.clj:394)", "query_processor$eval95267$process_query_and_save_execution_BANG___95276.invoke(query_processor.clj:387)", "query_processor$eval95312$process_query_and_save_with_max_results_constraints_BANG___95321$fn__95324.invoke(query_processor.clj:406)", "query_processor$eval95312$process_query_and_save_with_max_results_constraints_BANG___95321.invoke(query_processor.clj:399)", "api.dataset$run_query_async$fn__113755.invoke(dataset.clj:74)", "query_processor.streaming$streaming_response_STAR_$fn__99869$fn__99870.invoke(streaming.clj:166)", "query_processor.streaming$streaming_response_STAR_$fn__99869.invoke(streaming.clj:165)", "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:69)", "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:67)", "async.streaming_response$do_f_async$task__28914.invoke(streaming_response.clj:88)"],
	"card_id": null,
	"context": "ad-hoc",
	"error": "readStartArray can only be called when CurrentBSONType is ARRAY, not when CurrentBSONType is STRING.",
	"row_count": 0,
	"running_time": 0,
	"data": {
		"rows": [],
		"cols": []
	}
}
```
</details>

This looks like a separate issue though - is it @ranquild ?

I'll dig deeper in the meantime.

Reply:
maybe the reproduction e2e test is incorrect and needs to be fixed

Reply:
The problem is that the JSON query is getting extra `JSON.stringify`-ied when it's already JSON: https://github.com/metabase/metabase/blob/master/frontend/src/metabase/lib/engine.js#L89

![image](https://github.com/metabase/metabase/assets/6830683/9633bbd0-96ed-4c88-994d-c7f2fd804dbd)

It's a separate issue, but I guess I can fix it along this one - please let me know if I should report a separate issue.

Reply:
Related issues reported as #31121 & #31122